### PR TITLE
fix: GrouperAdmin Search in content model

### PR DESCRIPTION
**Pull request by mhsiddiqui**
## Related resources

* [Issue](https://github.com/django-cms/django-cms/issues/8277)

## Checklist

* [x] I have opened this pull request against ``main``
* [x] I have added or modified the tests when changing logic
* [x] I have followed [the conventional commits guidelines](https://www.conventionalcommits.org/) to add meaningful information into the changelog
* [x] I have read the [contribution guidelines ](https://github.com/django-cms/django-cms/blob/develop/CONTRIBUTING.rst) and I have joined the channel #pr-reviews on our [Discord Server](https://discord-pr-review-channel.django-cms.org) to find a “pr review buddy” who is going to review my pull request.

## Summary by Sourcery

Enable combined search across grouper and its related content model fields in the Django admin by extending the admin utils and adding comprehensive tests.

New Features:
- Allow GrouperAdmin to search by content model fields alongside grouper model fields
- Support Django admin search modifiers (^startswith and =exact) for both grouper and content fields

Enhancements:
- Override get_search_fields and get_search_results in admin utils to merge search results from the grouper model and its content model
- Refactor test helper createContentInstance to accept custom grouper instances and secret greetings

Tests:
- Add GrouperSearchTestCase with tests covering empty queries, model field searches, content field searches, modifier behaviors, and varied search_fields configurations

### CODE REVIEW

**Comment by fsbraun** [comment]
Hey @mhsiddiqui ! Thanks for this interesting PR. It is not an easy task - well done!

It is unfortunate that the Django admin does not expose the `construct_search` function inside `get_search_results`. This leads to a lot of code duplication and will be hard to maintain.

Can I make the following hacky suggestions:

* Let us try reusing the `super().get_search_results()` method. Once it should return the results for the grouper model, then for the content model. The overall result will be the union of both.
* Since `get_search_results` takes the query set as a parameter, we can pass it the content's admin_manager's `all()` qs, if we just could give it the right search fields, too.
* Here comes the hack: We override `get_search_fields` which takes the `request` argument. If that has a certain property, say, `_content_prefix`, it only returns the prefixed search fields (without the prefix for `get_search_result` to work), otherwise it only returns the unprefixed grouper search fields.

I agree that this is a bit hacky patching the `request`. What do you think?

**Comment by mhsiddiqui** [comment]
> Hey @mhsiddiqui ! Thanks for this interesting PR. It is not an easy task - well done!
> 
> It is unfortunate that the Django admin does not expose the `construct_search` function inside `get_search_results`. This leads to a lot of code duplication and will be hard to maintain.
> 
> Can I make the following hacky suggestions:
> 
> * Let us try reusing the `super().get_search_results()` method. Once it should return the results for the grouper model, then for the content model. The overall result will be the union of both.
> * Since `get_search_results` takes the query set as a parameter, we can pass it the content's admin_manager's `all()` qs, if we just could give it the right search fields, too.
> * Here comes the hack: We override `get_search_fields` which takes the `request` argument. If that has a certain property, say, `_content_prefix`, it only returns the prefixed search fields (without the prefix for `get_search_result` to work), otherwise it only returns the unprefixed grouper search fields.
> 
> I agree that this is a bit hacky patching the `request`. What do you think?

I tried approach with union of both queries by

File: cms/admin/utils.py [hunk 1]
```diff
@@ -696,6 +699,51 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+                grouper_search_fields.append(field_name)
+
+        if getattr(self, "_content_fields", False):
+            return content_search_fields
+
+        return grouper_search_fields
+
+    def get_search_results(self, request, queryset, search_term):
+        # Reset _content_fields flag
+        self._content_fields = False
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
        request._content_fields = False
```
I'm not entirely sure, but to make this more thread-safe, I'd add the flag to the `request` object. The admin object might be called from a different request while the flag is set.

File: cms/admin/utils.py [hunk 2]
```diff
@@ -696,6 +699,51 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+        content_search_fields = self.get_search_fields(request)
+        content_search_result, may_have_duplicate_content = self.content_model.admin_manager.none(), False
+        if content_search_fields:
+            content_search_result, may_have_duplicate_content = super().get_search_results(
+                request, content_queryset, search_term
+            )
+        search_result_from_content = queryset.filter(
+            id__in=content_search_result.values_list(f"{self.grouper_field_name}_id", flat=True)
+        )
+        return search_result_from_content, may_have_duplicate_content
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
        return search_result_from_content, False
```
The `id__in` filter will never have duplicated content.

File: cms/admin/utils.py [hunk 3]
```diff
@@ -696,6 +699,51 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+        content_search_fields = []
+        grouper_search_fields = []
+        for field_name in self.search_fields:
+            if field_name.startswith(f"{self.content_related_field}__") or field_name.startswith("content__"):
+                content_search_fields.append(re.sub(r"^([@^=]?)[^_]+__", r"\1", field_name))
+
+            else:
+                grouper_search_fields.append(field_name)
+
+        if getattr(self, "_content_fields", False):
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
        if getattr(request, "_content_fields", False):
```
I'd feel better if the parameter would carry the flag.

File: cms/admin/utils.py [hunk 4]
```diff
@@ -696,6 +699,51 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+
+        return grouper_search_result | search_result_from_content, (
+            may_have_duplicate_grouper & may_have_duplicate_content
+        )
+
+    def _get_content_search_result(self, request, queryset, search_term):
+        """Get search results from content model"""
+        content_queryset = self.content_model.admin_manager.all()
+        content_search_fields = self.get_search_fields(request)
+        content_search_result, may_have_duplicate_content = self.content_model.admin_manager.none(), False
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
        content_search_result, may_have_duplicate_content = self.content_model.admin_manager.none(), False
```
Can we put this into an else clause?

File: cms/admin/utils.py [hunk 5]
```diff
@@ -696,6 +699,51 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
             setattr(form._content_instance, self.grouper_field_name, obj)
             form._content_instance.save()
 
+    def get_search_fields(self, request):
+        """Return search fields for either grouper model or content model"""
+        content_search_fields = []
+        grouper_search_fields = []
+        for field_name in self.search_fields:
+            if field_name.startswith(f"{self.content_related_field}__") or field_name.startswith("content__"):
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
            if field_name.startswith(CONTENT_PREFIX):
```
For syntactical consistency, I would like to see the `CONTENT_PREFIX` here. This way all admin configurations behave the same.

File: cms/admin/utils.py [hunk 6]
```diff
@@ -696,6 +699,51 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+                content_search_fields.append(re.sub(r"^([@^=]?)[^_]+__", r"\1", field_name))
```
Comment (fsbraun) [inline cms/admin/utils.py]: This can be simplified to just remove `CONTENT_PREFIX`.

**Review by fsbraun — CHANGES_REQUESTED** [review]
Hi @mhsiddiqui ! This is looking good! I have only some minor comments (added). 

Can you check why tests fail on Django < 5.2?

**Comment by fsbraun** [comment]
@mhsiddiqui Once we get this merged, I'd port it back to the 5.0.2 release which the can be pushed out! Thanks!

File: cms/admin/utils.py [hunk 7]
```diff
@@ -696,6 +699,49 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
             setattr(form._content_instance, self.grouper_field_name, obj)
             form._content_instance.save()
 
+    def get_search_fields(self, request):
+        """Return search fields for either grouper model or content model"""
+        content_search_fields = []
+        grouper_search_fields = []
+        for field_name in self.search_fields:
+            if field_name.startswith(CONTENT_PREFIX):
+                content_search_fields.append(field_name.replace(CONTENT_PREFIX, ""))
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
                content_search_fields.append(field_name[len(CONTENT_PREFIX):]
```
The `CONTENT_PREFIX` could appear multiple times in the search field just accidentally. Let's make sure we only remove it once!

**Comment by mhsiddiqui** [comment]
> @mhsiddiqui Once we get this merged, I'd port it back to the 5.0.2 release which the can be pushed out! Thanks!

@fsbraun all changes recommended by you are done.

File: cms/admin/utils.py [hunk 8]
```diff
@@ -696,6 +696,49 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+        content_queryset = self.content_model.admin_manager.all()
+        content_search_fields = self.get_search_fields(request)
+        if content_search_fields:
+            content_search_result, __ = super().get_search_results(request, content_queryset, search_term)
+        else:
+            content_search_result = self.content_model.admin_manager.none()
+        search_result_from_content = queryset.filter(
+            id__in=content_search_result.values_list(f"{self.grouper_field_name}_id", flat=True)
+        )
+        return search_result_from_content, False
```
[the remark this answers is not in the document]
Comment (mhsiddiqui) [reply cms/admin/utils.py]: These changes require change in core Django functionality. We are currently avoiding any such change.

File: cms/admin/utils.py [hunk 9]
```diff
@@ -696,6 +696,49 @@
...
```
[the remark this answers is not in the document]
Comment (mhsiddiqui) [reply cms/admin/utils.py]: These changes require change in core Django functionality. We are currently avoiding any such change.

File: cms/admin/utils.py [hunk 10]
```diff
@@ -696,6 +696,49 @@
             setattr(form._content_instance, self.grouper_field_name, obj)
             form._content_instance.save()
 
+    def get_search_fields(self, request):
```
[the remark this answers is not in the document]
Comment (mhsiddiqui) [reply cms/admin/utils.py:699]: These changes require change in core Django functionality. We are currently avoiding any such change.

File: cms/admin/utils.py [hunk 11]
```diff
@@ -696,6 +696,48 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+            may_have_duplicate_grouper & may_have_duplicate_content
+        )
+
+    def _get_content_search_result(self, request, queryset, search_term):
+        """Get search results from content model"""
+        content_queryset = self.content_model.admin_manager.all()
+        if self.get_search_fields(request):
+            content_search_result, __ = super().get_search_results(request, content_queryset, search_term)
+        else:
+            content_search_result = self.content_model.admin_manager.none()
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
        try:
            # Set flag on request object to get the content search fields. `get_search_results` will call 
            # `get_search_fields` to get the content search fields.
            request._content_fields = True
            if self.get_search_fields(request):
                content_search_result, __ = super().get_search_results(request, content_queryset, search_term)
            else:
                content_serach_results = self.content_model.admin_manager.none()
        finally:
            request._content_fields = False
```
Setting the `_content_fields` flag in a try finally block will ensure that it is reset even if an exception occurs.

~~Also, the admin's `get_search_results` will evaluate the search fields - no need to go through the list twice.~~

File: cms/admin/utils.py [hunk 12]
```diff
@@ -696,6 +696,48 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+
+        return grouper_search_fields
+
+    def get_search_results(self, request, queryset, search_term):
+        # Reset _content_fields flag
+        request._content_fields = False
+        grouper_search_result, may_have_duplicate_grouper = super().get_search_results(request, queryset, search_term)
+
+        # Set _content_fields for get_search_fields to return content model search fields
+        request._content_fields = True
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
```
same

File: cms/admin/utils.py [hunk 13]
```diff
@@ -696,6 +696,48 @@ def save_model(self, request: HttpRequest, obj: models.Model, form: forms.Form,
...
+                grouper_search_fields.append(field_name)
+
+        if getattr(request, "_content_fields", False):
+            return content_search_fields
+
+        return grouper_search_fields
+
+    def get_search_results(self, request, queryset, search_term):
+        # Reset _content_fields flag
+        request._content_fields = False
```
Comment (fsbraun) [inline cms/admin/utils.py]:
```suggestion
```
This can be moved into `_get_content_search_result`

**Review by fsbraun — CHANGES_REQUESTED** [review]
Hi @mhsiddiqui ! This is turning out really nice! The only thing I'd like to ask you to look into is if we can encapsulate the request flag in a try finally block? This would make sure the flag is always removed - even if an exception was raised.

Out of curiosity: What was breaking the tests for Django < 5.2?

File: cms/admin/utils.py (hunk 11 above)
Comment (fsbraun) [reply cms/admin/utils.py]: I probably am wrong about the second point (crossed out now) here: If the search list is empty the `super().get_search_results()` call might return **all** content objects - we only want additional ones.

**Comment by mhsiddiqui** [comment]
> Hi @mhsiddiqui ! This is turning out really nice! The only thing I'd like to ask you to look into is if we can encapsulate the request flag in a try finally block? This would make sure the flag is always removed - even if an exception was raised.
> 
> Out of curiosity: What was breaking the tests for Django < 5.2?

@fsbraun tests were failing due to some unused imports which I actually added for my previous logic but forgot to remove when logic was changed.

**Review by fsbraun — APPROVED** [review]
🎉
